howtos/diffs/scheduled-sampling.diff

```diff
@@ -0,0 +1,125 @@
++      x = jnp.where(
++          jax.random.bernoulli(
++              bernoulli_rng, p=sample_probability, shape=(batch_size, 1)),
++          last_prediction, x)
+       lstm_state, y = lstm_cell(lstm_state, x)
+       logits = projection(y)
+       predicted_tokens = jax.random.categorical(categorical_rng, logits)
+@@ -223,7 +225,7 @@ class Seq2seq(nn.Module):
+   def apply(self,
+             encoder_inputs,
+             decoder_inputs,
+-            teacher_force=True,
++            sample_probability=0.0,
+             eos_id=1,
+             hidden_size=512):
+     """Run the seq2seq model.
+@@ -234,14 +236,15 @@ class Seq2seq(nn.Module):
+         `[batch_size, max(encoder_input_lengths), vocab_size]`.
+       decoder_inputs: padded batch of expected decoded sequences for teacher
+         forcing, shaped `[batch_size, max(decoder_inputs_length), vocab_size]`.
+-        When sampling (i.e., `teacher_force = False`), the initial time step is
+-        forced into the model and samples are used for the following inputs. The
+-        second dimension of this tensor determines how many steps will be
+-        decoded, regardless of the value of `teacher_force`.
+-      teacher_force: bool, whether to use `decoder_inputs` as input to the
+-        decoder at every step. If False, only the first input is used, followed
+-        by samples taken from the previous output logits.
+-      eos_id: int, the token signaling when the end of a sequence is reached.
++        The initial time step is forced into the model and samples are used for
++        the following inputs as determined by `sample_probability`. The second
++        dimension of this tensor determines how many steps will be
++        decoded, regardless of the value of `sample_probability`.
++      sample_probability: float in [0, 1], the probability of using the previous
++        sample as the next input instead of the value in `decoder_inputs` when
++        sampling. A value of 0 is equivalent to teacher forcing and 1 indicates
++        full sampling.
++      eos_id: int, the token signalling when the end of a sequence is reached.
+       hidden_size: int, the number of hidden dimensions in the encoder and
+         decoder LSTMs.
+     Returns:
+@@ -255,7 +258,7 @@ class Seq2seq(nn.Module):
+     logits, predictions = decoder(
+         init_decoder_state,
+         decoder_inputs[:, :-1],
+-        teacher_force=teacher_force)
++        sample_probability=sample_probability)
+ 
+     return logits, predictions
+ 
+@@ -325,14 +328,15 @@ def compute_metrics(logits, labels):
+ 
+ 
+ @jax.jit
+-def train_step(optimizer, batch, rng):
++def train_step(optimizer, batch, sample_probability, rng):
+   """Train one step."""
+   labels = batch['answer'][:, 1:]  # remove '=' start token
+ 
+   def loss_fn(model):
+     """Compute cross-entropy loss."""
+-    with nn.stochastic(rng):
+-      logits, _ = model(batch['query'], batch['answer'])
++    logits, _ = model(
```

Comment:
Is `with nn.stochastic(rng)` no longer necessary? That would surprise me...

Reply:
(Or if so, it should be removed from the base example. I suppose we could do that in a separate commit)

Reply:
Fixed and tested.